### PR TITLE
Feature/spark3 module

### PR DESCRIPTION
**Pull request by kevinwallimann**
1. Added 5 new modules
- compatibility-api
- compatibility-provider
- compatibility_spark-2
- compatibility_spark-3
- hyperdrive-release_spark-3
The hyperdrive-release module was renamed to hyperdrive-release_spark-2

compatibility-api is parent to the other compatibility* modules. It defines the API that is implemented by all three child modules. `compatibility_spark-2` contains only spark 2 dependencies, while `compatibility_spark-3` contains only spark 3 dependencies. Both modules implement the API under the same namespaces. `compatibility-provider` implements the API by forwarding to either `compatibility_spark-2` or `compatibility_spark-3`, controlled by the maven profile.
Any other module will declare `compatibility-provider` as a dependency.

2. Added new profiles
- spark-2
- spark-3

3. Default build `mvn install` will build with scala 2.12 / spark 2.4

4. The uber-jar will be available in three variants
- hyperdrive-release_spark-2_2.11
- hyperdrive-release_spark-2_2.12
- hyperdrive-release_spark-3_2.12

The deployment script was tested with a local nexus instance. 
The uber jars were tested locally with Spark 2.4.7 / Scala 2.11, Spark 2.4.2 / Scala 2.12 and Spark 3.0.1 / Scala 2.12

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -154,6 +171,19 @@
                 <scala.version>${scala_2.12.version}</scala.version>
             </properties>
         </profile>
+        <profile>
+            <id>spark-3</id>
+            <properties>
+                <spark.compat.version>spark-3</spark.compat.version>
+                <spark.version>${spark_3.version}</spark.version>
+                <spark.sql.mongo.version>${spark_3.sql.mongo.version}</spark.sql.mongo.version>
```
Comment (Zejnilovic) [inline pom.xml:193]: when you activate the build with profile `spark-3` you do not deactivate the `spark-2` modules, which means you use spark3 versions of these libraries for the build of spark2 modules as well. Right?

Comment (kevinwallimann) [reply pom.xml:193]: Ah, I notice that I didn't test the spark 2.4 / scala 2.12 combination. It won't work in the current configuration.

The idea is that all modules except `compatibility_spark-2`, `compatibility_spark-3`, `hyperdrive-release_spark-2` and `hyperdrive-release_spark-3` are compatible for both spark 2.4 and 3.0. So e.g. if you only depend on the API module, you don't have to worry about spark 2.4 vs 3.0. `compatibility_spark-2` always uses spark 2 dependencies, even if profile `spark-3` is activated, and `hyperdrive-release_spark-2` should always use the `compatibility_spark-2` module. This last part I forgot to implement, I'll fix this

File: pom.xml [hunk 2]
```diff
@@ -171,6 +169,22 @@
                 <scala.version>${scala_2.12.version}</scala.version>
             </properties>
         </profile>
+        <profile>
+            <id>spark-2</id>
+            <activation>
+                <activeByDefault>true</activeByDefault>
```
Comment (Zejnilovic) [inline pom.xml:175]: I think that if you turn on profile `deploy` without the `spark-2` or `spark-3`, `spark-2` will be turned off. right? Or do I misunderstand?

Comment (kevinwallimann) [reply pom.xml:175]: Yes that's correct. My main motivation for the `activeByDefault` flag was to ensure `mvn clean package` works without any profiles explicitly set

**Review by Zejnilovic — DISMISSED** [review]

**Review by jozefbakus — APPROVED** [review]

**Pull request merged**